Scripts and tools: gitian-build.py improvements and corrections

hebasto opened the pull request:
1. The Docker does not depend on `apt-cacher-ng` package. Ref: #14002.

2. Do not try to install the Docker if `docker.service` is detected on the system (e.g., the Docker was installed manually). Fix https://github.com/bitcoin/bitcoin/pull/13623#issuecomment-405684241 by **Sjors**.

3. Prevent the setting of more than one environment variable for the `gitian-builder` (an alternative to #13999). E.g., USE_LXC being set shadows USE_DOCKER; for details see [`gitian-builder/libexec/make-clean-vm`](https://github.com/devrandom/gitian-builder/blob/93a62c7d7d018c66c02a19bac3d751144043cfec/libexec/make-clean-vm#L7):
```sh
VMSW=KVM
if [ -n "$USE_LXC" ]; then
    VMSW=LXC
elif [ -n "$USE_VBOX" ]; then
    VMSW=VBOX
elif [ -n "$USE_DOCKER" ]; then
    VMSW=DOCKER
fi
```
4. The [`gitian-builder/bin/gverify`](https://github.com/devrandom/gitian-builder/blob/master/bin/gverify) script returns the exit code 1 if a signature verification ends with 'BAD SIGNATURE' or 'MISMATCH' by design. This PR allows to see the verification results for all signatures without a premature fail of the `gitian-build.py` script. Ref: #14014.

ken2812221 reviewed (approved):
utACK c45cf0ba56d1405d8c6a87d899a8390c5a6e4b0b. Some nits and please squash.

MarcoFalke commented on contrib/gitian-build.py:
```diff
@@ -143,9 +143,8 @@ def main():
-    parser.add_argument('-k', '--kvm', action='store_true', dest='kvm', help='Use KVM instead of LXC')
-    parser.add_argument('-d', '--docker', action='store_true', dest='docker', help='Use Docker instead of LXC')
-    parser.add_argument('-S', '--setup', action='store_true', dest='setup', help='Set up the Gitian building environment. Uses LXC. If you want to use KVM, use the --kvm option. Only works on Debian-based systems (Ubuntu, Debian)')
+    parser.add_argument('-V', '--virtualization', dest='virtualization', default='lxc', help='Specify virtualization technology to use: lxc for LXC, kvm for KVM, docker for Docker. Default is %(default)s')
```
Just noting that external documentation would have to be updated to reflect this change.

hebasto replied:
All related documentation which describes this script options is located in the https://github.com/bitcoin-core/docs repository. Therefore, IMO, there is nothing to add to this PR regarding docs.
Or did I miss something?

I'll submit an appropriate PR to the https://github.com/bitcoin-core/docs repository in a timely manner.

MarcoFalke replied:
This can be achieved without a breaking change via

```
args.lxc = not args.kvm and not args.docker

hebasto commented:
Rebased.

ken2812221 commented:
re-utACK be88ede

hebasto committed:
Fix gitian-build.py --verify option

The gitian-builder/bin/gverify script returns the exit code 1 if a
signature verification ends with 'BAD SIGNATURE' or 'MISMATCH' by
design. This commit allows to see the verification results for all
signatures without a premature fail of the gitian-build.py script.

hebasto commented:
@fanquake #14014 has been combined into this PR. Would you mind reviewing it?

charlesrocket commented:
we could drop https://github.com/bitcoin/bitcoin/blob/dbd217be69ab6453f0adeb52cbacdace78406a45/contrib/gitian-build.py#L215-L218 when/if #14539 is merged

hebasto commented:
@fanquake May I ask you to restart travis against this PR?

ken2812221 commented:
utACK dbd217be69ab6453f0adeb52cbacdace78406a45
travis restarted

MarcoFalke commented on contrib/gitian-build.py:
```diff
@@ -118,15 +118,15 @@ def verify():
+    subprocess.call(['bin/gverify', '-v', '-d', '../gitian.sigs/', '-r', args.version+'-win-signed', '../bitcoin/contrib/gitian-descriptors/gitian-win-signer.yml'])
     print('\nVerifying v'+args.version+' Signed MacOS\n')
-    subprocess.check_call(['bin/gverify', '-v', '-d', '../gitian.sigs/', '-r', args.version+'-osx-signed', '../bitcoin/contrib/gitian-descriptors/gitian-osx-signer.yml'])
+    subprocess.call(['bin/gverify', '-v', '-d', '../gitian.sigs/', '-r', args.version+'-osx-signed', '../bitcoin/contrib/gitian-descriptors/gitian-osx-signer.yml'])
```
Why should this not fail if there is a mismatch?

hebasto replied:
The [`gitian-builder/bin/gverify`](https://github.com/devrandom/gitian-builder/blob/master/bin/gverify) script returns the exit code `1` if a signature verification ends with 'BAD SIGNATURE' or 'MISMATCH' by design.

From the Python docs:
- [`subprocess.check_call()`](https://docs.python.org/3.8/library/subprocess.html#subprocess.check_call):
> If the return code was zero then return, otherwise raise CalledProcessError.

- [`subprocess.call()`](https://docs.python.org/3.8/library/subprocess.html#subprocess.call):
>  Wait for command to complete, then return the returncode attribute.

MarcoFalke replied:
I understand that by this change the return code is ignored, but I'd like to understand why.

If the signatures are bad or the hashes mismatch, the verification should fail.

MarcoFalke replied:
Fine, if you want to print the result for all signatures, but please don't change the behaviour (fail on non-zero exit code) of the script.

MarcoFalke commented:
utACK be88ede292337f40617ccc7f85f591f2a234cc26

hebasto commented:
@MarcoFalke Thank you for your review.
> Fine, if you want to print the result for all signatures, but please don't change the behaviour (fail on non-zero exit code) of the script.

Fixed. Would you mind re-reviewing?

hebasto committed:
Fix Docker related issues for gitian-build.py

The Docker does not depend on apt-cacher-ng package.
Do not try to install the Docker if docker.service is detected on the
system (e.g., the Docker was installed manually).
Also small style corrections were applied.

hebasto committed:
Set/unset USE_LXC, USE_VBOX, USE_DOCKER explicitly

This prevents the setting of more than one environment variable for the
gitian-builder (e.g., USE_LXC being set shadows USE_DOCKER; for details
see gitian-builder/libexec/make-clean-vm).

hebasto commented:
Rebased.

@MarcoFalke 
> This can be achieved without a breaking change via
`args.lxc = not args.kvm and not args.docker`

Dropped new `--virtualization` option. So, no breaking changes now.

Also "do not warn about macOS build on setup" feature dropped in favor of  #15236.

Commit messages have been brushed.

hebasto commented:
@Sjors 
Would you mind reviewing this PR regarding [your comment](https://github.com/bitcoin/bitcoin/pull/13623#issuecomment-405684241)?
